[DEV-329] Metrics and analytics for bot commands

wozniakjan opened the pull request:
## Issue ID(s):
https://ddevhq.atlassian.net/browse/DEV-329

We would like to have some insight into how many times the preview bot commands have been invoked, similar to how we measure CLI or UI or API calls. This tracks all three preview bot commands and groups them into few error / success groups.

There are two sinks for these metrics:
* prometheus `ServiceMonitor`
* segment / amplitude analytics

## Tests:
<!-- Tests introduced by this PR, manual steps to validate the fix, or an explanation for why no tests are needed. -->

## Release/Deployment notes:
<!-- Are there ramifications to other code? Does anything have to be done on deployment? -->
example for prometheus metrics:
```
$ curl --silent localhost:8080/metrics  | grep bot
# HELP bot_create_preview_site_total Total number of create preview site command invocations
# TYPE bot_create_preview_site_total counter
bot_create_preview_site_total{bot="github-operator",status="success"} 1
# HELP bot_help_preview_site_total Total number of help command invocations
# TYPE bot_help_preview_site_total counter
bot_help_preview_site_total{action="command",bot="github-operator"} 1
bot_help_preview_site_total{action="pr",bot="github-operator"} 1
```

example for analytics avent as taken from segment
```go
client.Track(&analytics.Track{
  UserId: "wozniak.jan@gmail.com",      // the user email as we receive it from github / gitlab
  Event: "/ddev-live-help",             // the command they executed with the bot
  Properties: map[string]interface{}{
    "bot": "github-operator",           // github or gitlab bot
    "result": "success",                // whether the bot succe
    "type": "pr",                       // either "pr" or "command", whether the action was triggered by the user typing the command or whether it was auto triggered by PR open / close
  },
})
```

wozniakjan commented on pkg/metrics.go:
```diff
@@ -0,0 +1,50 @@
+	CreatePreviewSiteCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "bot_create_preview_site_total",
+		Help: "Total number of create preview site command invocations",
+	}, []string{"bot", "status"})
```
`"bot"` means the type of the bot, we run two, one as part of gitlab-webhook-server and second under github-operator.
`"status"` is one of the five statuses described below, to give us a better idea of how frequently our users struggle with the bot.

wozniakjan commented on pkg/metrics.go:
```diff
@@ -0,0 +1,50 @@
+	DeletePreviewSiteCounter = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "bot_delete_preview_site_total",
+		Help: "Total number of delete preview site command invocations",
+	}, []string{"bot", "action", "status"})
```
`"action"` can be either `PR`/`MR` or `command`. Some commands are invoked automatically when the PR is created or closed, all can be invoked by a user entering a command as a comment. Wanted to distinguish between these two cases.

wozniakjan commented on pkg/bot.go:
```diff
@@ -464,9 +475,14 @@ func (b *bot) hasPreviewSiteCapability(email, org string) (bool, error) {
 }
 
 func (b *bot) previewSite(args ResponseRequest) string {
+	if args.Email == "" {
```
moving this higher up the call to save CPU some cycles, when user has no public email, we can stop there and let them know.

drekle commented on pkg/bot.go:
```diff
@@ -464,9 +475,14 @@ func (b *bot) hasPreviewSiteCapability(email, org string) (bool, error) {
 	list, err := b.sisLister.List(labels.Everything())
 	if err != nil {
 		klog.Errorf("failed listing SiteImageSource: %v", err)
+		CreatePreviewSiteCounter.WithLabelValues(b.annotation, metricLabelPreviewFailed).Inc()
```
Not sure it matters, however it may be easier to alert on separated.  Also prometheus advise against labels generally, however I think that advice is more around dynamic things you cannot measure like the name of pods.

![image](https://user-images.githubusercontent.com/1883554/104052192-91f36300-51a6-11eb-947b-991c8a6b94bc.png)
![image](https://user-images.githubusercontent.com/1883554/104052209-991a7100-51a6-11eb-8bcf-bce113f078b9.png)

However also an alert doing the same:
![image](https://user-images.githubusercontent.com/1883554/104052243-a6376000-51a6-11eb-9590-17277f11682a.png)

wozniakjan replied:
> any reason to not have a separate timeseries for each status?

no strong reason really, I loosely modelled this after `htttp_requests_total`. That metric also has `status` in one of its labels and my thinking was "yeah, that sound like similar usage". I think both will behave similarly, in both cases, there is going to be own time series, likely going to consume a similar amount of memory. Some queries are going to be harder, with a label, some easier.

> Also prometheus advise against labels generally, however I think that advice is more around dynamic things you cannot measure like the name of pods.

https://prometheus.io/docs/practices/naming/#labels, that is how I am interpreting their advice as well. This case, when I have a small bounded number of labels, I think is in correlation with general usage and practices.

Weakly leaning towards keeping it as a label but happy to revert to own TS.

drekle reviewed (approved):
lgtm

wozniakjan commented:
I was about to merge this PR and integrate it to the operators when I noticed that https://ddevhq.atlassian.net/browse/DEV-329 has a specific requirement to be able to filter by customer account.

> A report is generated that tracks how many times a specific customer account has triggered Preview bots, over a specified time period

@drekle, I understand this makes the growth exponential and is against general recommendations around prometheus. Do we go with it and expect to keep prometheus well-sized or do we shove this metric elsewhere?

drekle commented on pkg/bot.go:
```diff
@@ -149,24 +160,25 @@ func authAdminAPI(k *kubernetes.Clientset, authSVCLabels string) (pbAdmin.Admini
 
-func InitBot(kubeconfig *restclient.Config, annotation, siteSuffix string, authSVCLabels string, stopCh <-chan struct{}) (Bot, error) {
-	scs, err := siteclientset.NewForConfig(kubeconfig)
+func InitBot(bc Config) (Bot, error) {
```
Is this called by another service?  Are we mounting the AnalyticsKey via secret?

wozniakjan replied:
this is called by github-operator and gitlab-webhook-server. Yes, my plan was to mount the key via `SealedSecret` in a state repo, it has to be different for prod and non-prod.

drekle replied:
Thanks.  Theres a bit of oddness required in how the CLI bakes this in I am sure you noticed, perhaps required for CLI's distributed to our users.

drekle commented on pkg/metrics.go:
```diff
@@ -0,0 +1,160 @@
+	}
+}
+
+func (a *analytics) CreatePreviewSite(user, result string) {
```
_Perhaps just thinking out loud here._

Not a huge fan of these function names but struggle to think of one that resolves my conflict with them.  They are named in such a way to perform an action whereas they only record metrics, suppose that it could be gathered from the structure in which the functions belong.  `IncCreatePreview`  `ObserveCreatePreview`  Or could have seen a singular `Observe` function which could take a:
```
func (a *analytics) Observe(cmd BotCommand, user string, result string){
```
if we typed them:
```
type BotCommand string
const PreviewSite BotCommand = "/ddev-live-preview-site"
```
https://github.com/drud/repo-chat-bot/blob/main/pkg/responses.go#L46

However from my understanding PreviewSite is just a string and we don't really need to type them also.

wozniakjan replied:
I like the `Observe` prefix and maybe, even more, the singular `Observer()` function, will see how these two look and pick one that I like better.

wozniakjan replied:
so because these functions take slightly different arguments, I decided to go with the `Observe` prefix.

drekle reviewed (approved):
lgtm

wozniakjan commented:
@pgalligan80126 @rfay, I added you here so you have some visibility into how I have implemented the feed to segment.com, no need to do a code review :)

event example:
```go
client.Track(&analytics.Track{
  UserId: "wozniak.jan@gmail.com",      // the user email as we receive it from github / gitlab
  Event: "/ddev-live-help",             // the command they executed with the bot
  Properties: map[string]interface{}{
    "bot": "github-operator",           // github or gitlab bot
    "result": "success",                // whether the bot succe
    "type": "pr",                       // either "pr" or "command", whether the action was triggered by the user typing the command or whether it was auto triggered by PR open / close
  },
})
```

pgalligan80126 commented:
will the output be a Grafana report or segment?  @wozniakjan

wozniakjan commented:
> will the output be a Grafana report or segment? @wozniakjan

for dev environment it is segment https://app.segment.com/drud/sources/ddev_live_bot_dev/debugger, for prod it will be segment https://app.segment.com/drud/sources/ddev_live_bot/debugger (has currently no data because it's waiting for a release). Prod has also been connected to amplitude by @unn.
